server/pulp/plugins/file/model_distributor.py

```diff
@@ -0,0 +1,249 @@
+        :rtype:  dict
+        """
+        raise NotImplementedError()
+
+    def distributor_removed(self, repo, config):
+        """
+        Delete the published files from our filesystem
+
+        Please also see the superclass method definition for more documentation on this method.
+
+        :param repo: metadata describing the repository
+        :type  repo: pulp.server.db.model.Repository
+
+        :param config: plugin configuration
+        :type  config: pulp.plugins.config.PluginCallConfiguration
+        """
+        self.unpublish_repo(repo, config)
+
+    def publish_repo(self, repo, publish_conduit, config):
+        """
+        Publish the repository.
+
+        :param repo:            metadata describing the repo
+        :type  repo:            pulp.server.db.model.Repository
+        :param publish_conduit: The conduit for publishing a repo
+        :type  publish_conduit: pulp.plugins.conduits.repo_publish.RepoPublishConduit
+        :param config:          plugin configuration
+        :type  config:          pulp.plugins.config.PluginConfiguration
+        :param config_conduit: Configuration Conduit;
+        :type config_conduit: pulp.plugins.conduits.repo_validate.RepoConfigConduit
+        :return:                report describing the publish operation
+        :rtype:                 pulp.plugins.model.PublishReport
+        """
+        progress_report = FilePublishProgressReport(publish_conduit)
+        _logger.info(_('Beginning publish for repository <%(repo)s>') % {'repo': repo.id})
+
+        try:
+            progress_report.state = progress_report.STATE_IN_PROGRESS
+            repo_model = model.Repository.objects.get_repo_or_missing_resource(repo.id)
+            units = repo_controller.find_repo_content_units(repo_model, yield_content_unit=True)
+
+            # Set up an empty build_dir
+            working_dir = common_utils.get_working_directory()
+            build_dir = os.path.join(working_dir, BUILD_DIRNAME)
+            os.makedirs(build_dir)
+
+            self.initialize_metadata(build_dir)
+
+            try:
+                # process each unit
+                for unit in units:
+                    links_to_create = self.get_paths_for_unit(unit)
+                    self._symlink_unit(build_dir, unit, links_to_create)
+                    self.publish_metadata_for_unit(unit)
+            finally:
+                # Finalize the processing
+                self.finalize_metadata()
+
+            # Let's unpublish, and then republish
+            self.unpublish_repo(repo_model, config)
+
+            hosting_locations = self.get_hosting_locations(repo_model, config)
+            for location in hosting_locations:
+                shutil.copytree(build_dir, location, symlinks=True)
+
+            self.post_repo_publish(repo_model, config)
+
+            # Clean up our build_dir
+            common_utils.delete_working_directory()
```

Comment:
We probably don't need states like this anymore. Working dirs are automatically deleted upon task success/failure and also worker start.